euphonic/plot.py

```diff
@@ -33,37 +33,67 @@ def plot_1d_to_axis(spectra: Union[Spectrum1D, Spectrum1DCollection],
         Spectrum1D or Spectrum1DCollection to plot
     ax
         Matplotlib axes to which spectra will be drawn
+    labels
+        The label to be applied to each line in spectra, must be the same
+        as the number of lines. If this is None, the label(s) contained in
+        spectra.metadata 'label' (Spectrum1D) or 'labels'
+        (Spectrum1DCollection) will be used. To disable labelling for a
+        specific line, pass an empty string.
     **mplargs
         Keyword arguments passed to Axes.plot
     """
 
     if isinstance(spectra, Spectrum1D):
         return plot_1d_to_axis(Spectrum1DCollection.from_spectra([spectra]),
-                             ax=ax, **mplargs)
+                               ax=ax, labels=labels, **mplargs)
 
     try:
         assert isinstance(spectra, Spectrum1DCollection)
     except AssertionError:
         raise TypeError("spectra should be a Spectrum1D or "
                         "Spectrum1DCollection")
 
+    if labels is not None and len(labels) != len(spectra):
+        raise ValueError(
+            f"The length of labels (got {len(labels)}) should be the "
+            f"same as the number of lines to plot (got {len(spectra)})")
+
     # Find where there are two identical x_data points in a row
     breakpoints = (np.where(spectra.x_data.magnitude[:-1]
                             == spectra.x_data.magnitude[1:])[0]
                    + 1).tolist()
     breakpoints = [0] + breakpoints + [None]
 
-    for spectrum in spectra:
-        # Plot each line in segments, keeping colour consistent
+    for i, spectrum in enumerate(spectra):
+
+        # Plot each line in segments
         for x0, x1 in zip(breakpoints[:-1], breakpoints[1:]):
+            # Keep colour consistent across segments
             if x0 == 0:
                 color = None
             else:
                 color = p[-1].get_color()
 
+            # Optionally add legend label to the first segment
+            label = None
+            if x0 == 0:
+                if labels is None:
+                    if spectrum.metadata is not None:
+                        try:
+                            label = spectrum.metadata['label']
+                        except KeyError:
+                            pass
+                else:
+                    label = labels[i]
+
             p = ax.plot(spectrum.get_bin_centres().magnitude[x0:x1],
                         spectrum.y_data.magnitude[x0:x1],
-                        color=color, **mplargs)
+                        color=color, label=label, **mplargs)
+
+    # Update legend if it exists, in case new labels have been added
```

Comment:
It depends, as this is a user-facing function, do we want them to have to manually redraw the legend after they've added a line? I would've thought if they're adding a line with a label it would be expected that the legend will be updated to reflect that automatically.

Reply:
The usual design pattern for matplotlib scripts is not to draw the legend until you're done plotting. I suppose this is nice for interactive use? I can't think of a case where it would be a _problem_, except that it might make debugging confusing if people don't know the feature is there.